From 9cb0fa530fe3e01543c3f9467765820fbc485711 Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Date: Wed, 17 Jun 2015 18:37:49 +0300
Subject: [PATCH 169/452] dpaa2-eth: Support fewer DPIOs than CPUs

The previous DPIO functions would transparently choose a (perhaps
non-affine) CPU if the required CPU was not available. Now that their API
contract is enforced, we must make an explicit request for *any* DPIO if
the request for an *affine* DPIO has failed.

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Change-Id: Ib08047ffa33518993b1ffa4671d0d4f36d6793d0
Reviewed-on: http://git.am.freescale.net:8181/38320
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Roy Pledge <roy.pledge@freescale.com>
Reviewed-by: Stuart Yoder <stuart.yoder@freescale.com>
[Xulin: Original patch taken from FSL LS2085 SDK EAR6.0,
LS2085A-SDK-SOURCE-20160304-yocto.iso]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c |   18 ++++++++++++++----
 1 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c b/drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c
index d3c71e6..8bb1169 100644
--- a/drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c
+++ b/drivers/staging/fsl-dpaa2/ethernet/ldpaa_eth.c
@@ -1253,10 +1253,20 @@ static int __cold __ldpaa_dpio_setup(struct ldpaa_eth_priv *priv,
 		/* Register the new context */
 		err = dpaa_io_service_register(NULL, nctx);
 		if (unlikely(err)) {
-			netdev_err(priv->net_dev,
-				   "Rx notifications register failed\n");
-			nctx->cb = NULL;
-			goto err_service_reg;
+			dev_info_once(priv->net_dev->dev.parent,
+				     "Could not get (some) affine DPIO(s), probably there are not enough of them in the DPL\n");
+			/* Try to get *any* portal, not necessarily affine to
+			 * the requested cpu. This might be the case if there
+			 * are fewer DPIO objects in the container than CPUs.
+			 */
+			nctx->desired_cpu = -1;
+			err = dpaa_io_service_register(NULL, nctx);
+			if (unlikely(err)) {
+				dev_err(priv->net_dev->dev.parent,
+					"Could not get any DPIO!\n");
+				nctx->cb = NULL;
+				goto err_service_reg;
+			}
 		}
 	}
 
-- 
1.7.5.4

